src/utils/validators.py

```
@@ -14,19 +15,26 @@ def json_list_parser(model_cls: type[BaseModel]):
-                    f"got {type(parsed).__name__}: {parsed}"
-                )
-            return [model_cls.model_validate(item) for item in parsed]
-        return v
+                    f"Failed to parse JSON (auto-repair failed): {e}"
+                ) from e
+
+        if not isinstance(parsed, list):
+            raise ValueError(
+                f"Expected JSON array for {model_cls.__name__}, "
+                f"got {type(parsed).__name__}: {parsed}"
+            )
```

Ruff (0.14.5)

Prefer `TypeError` exception for invalid type

(TRY004)